### PR TITLE
test: Make HTTP requests instead of calling endpoints in-process

### PR DESCRIPTION
**Pull request by badr4php**
Calling functions in-process does not test the deployed service. The tests in the starter app should set the precedent for how we test our services.

Related to Jira: [ENG-1451]

[ENG-1451]: https://comicrelief.atlassian.net/browse/ENG-1451?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by seb-cr — CHANGES_REQUESTED** [review]
This is still calling the `HelloAction` function in the tests, instead of making a HTTP request to `serverless-offline`. Sorry I wasn't very clear on the Jira ticket.

Essentially we need an `axios` call to `process.env.BASE_URL`. For an example see the `serverless-giftaid` tests: https://github.com/comicrelief/serverless-giftaid/blob/master/tests/feature/submit/empty.feature.js

**Comment by badr4php** [comment]
@seb-cr   I used  `axios` to call `hello action` but the response's headers  be different, to be honest I don't know if I should to fix the response or just edit the test assertions .
The new headers is :     {
      'content-type': 'application/json; charset=utf-8',
      'access-control-allow-credentials': 'true',
      vary: 'origin',
      'access-control-expose-headers': 'WWW-Authenticate,Server-Authorization',
      'cache-control': 'no-cache',
      'content-length': '44',
      'accept-ranges': 'bytes',
      date: 'Thu, 21 Apr 2022 16:38:43 GMT',
      connection: 'close'
    }

**Comment by seb-cr** [comment]
> I used `axios` to call `hello action`

Thanks @badr4php, that's looking good now 🙂

> but the response's headers be different, to be honest I don't know if I should to fix the response or just edit the test assertions .

Interesting, you're right, the headers are missing `access-control-allow-origin`. Ultimately what we want is for the headers to match what AWS would give us when we make a request to a deployment of the service. It might be that the old assertions are wrong if AWS API Gateway modifies them. Let me have a look at this tomorrow and get back to you!

**Comment by corinja** [comment]
Unless anyone stops me, I'll deploy this to the AWS sandbox account and have a look at the response headers from there to try to resolve this issue about the `Access-Control-Allow-Origin` header in the tests @seb-cr @pauldolden

**Comment by seb-cr** [comment]
> Unless anyone stops me, I'll deploy this to the AWS sandbox account and have a look at the response headers from there to try to resolve this issue about the `Access-Control-Allow-Origin` header in the tests

Sounds good. I don't remember what the outcome was when I was looking at this – almost a year ago 😬

**Comment by corinja** [comment]
Confirmed: headers are different from a deployed service in AWS to the `serverless-offline` version.

Sandbox deployment headers:

image

Local sls-offline headers:

image

**Comment by seb-cr** [comment]
Thanks Corin. So it looks like API Gateway passes the headers back as-is from Lambda, whereas `serverless-offline` is doing several things to the headers to make them more "correct". It's added the encoding to `Content-Type`, and I have a suspicion that `Vary: origin` is equivalent to `Access-Control-Allow-Origin: *` (though my knowledge of CORS is limited!)

I wonder if this is still the case in the latest version of `serverless-offline`. Would need a bit of work to find out because it requires `serverless@3`, so we'd need to do that update first...

**Comment by corinja** [comment]
Getting an error on this when merging in the changes to bring it to `Node` 14

**Comment by corinja** [comment]
`serverless-framework` v3 with the latest version of `serverless-offline` works for this app with one small change to the way they `apiKey` argument is passed in (see https://www.serverless.com/framework/docs/guides/upgrading-v3#upgrading-to-v3 `CLI commands and options`) but it doesn't fix the CORS issue.

Running `sls offline` and then issuing

```bash
curl -v -X OPTIONS http://localhost:3001/hello 
```

gives

```bash
*   Trying 127.0.0.1:3001...
* Connected to localhost (127.0.0.1) port 3001 (#0)
> OPTIONS /hello HTTP/1.1
> Host: localhost:3001
> User-Agent: curl/7.84.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< content-type: application/json; charset=utf-8
< cache-control: no-cache
< content-length: 107
< Date: Mon, 06 Mar 2023 16:46:56 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< 
* Connection #0 to host localhost left intact
{"statusCode":404,"error":"Not Found","message":"CORS error: Missing Access-Control-Request-Method header"}%   
```

**Comment by seb-cr** [comment]
Can you try the same thing with a `GET` request? I don't understand why the response status code is 404... You might also need to add the `Origin` header (`-H 'Origin: https://example.com'`).

**Comment by corinja** [comment]
That works fine:

```bash
curl -H 'Origin: https://example.com' http://localhost:3001/hello 
{"data":{"response":"Hello"},"message":"ok"}
```

**Comment by corinja** [comment]
Tests also pass now

```bash
$ USE_SERVERLESS_OFFLINE=1 yarn test:feat
$ yarn test --selectProjects FEAT
$ jest --colors --selectProjects FEAT
Running one project: FEAT
 PASS   FEAT  tests/feat/hello.spec.js
  feat.Hello
    ✓ should return a successful status (13 ms)
    ✓ should return CORS headers (1 ms)
    ✓ should respond with "Hello" (1 ms)

{ level: 'info', message: 'metric - lambda.statusCode - 200' }
{ level: 'info', message: 'metric - lambda.statusCode - 200' }
{ level: 'info', message: 'metric - lambda.statusCode - 200' }
Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
```
I propose to close this PR and submit a new one to upgrade to serverless framework v3 and then a new PR for the changes included here.

**Review by seb-cr — COMMENTED** [review]
Couple more thoughts in the comments, otherwise I'm happy with this apart from the two points below.

- Tests fail when I run `yarn test:offline` locally without setting `BASE_URL` in my `.env`. The `test:offline` command should override the env as we have a special case for it in [`tests/setup.js`](https://github.com/comicrelief/serverless-starter-app/blob/a2a4770d1d686c0f72f441f0bc739b4bd9ae9763/tests/setup.js#L6), however it looks like this sets `SERVICE_URL` instead of `BASE_URL`. Suggesting we change all instances to `SERVICE_BASE_URL` for best clarity (it's a base URL, and it's for the service).

- We can now delete `tests/lib/lambda-promisifier.js`and `tests/lib/lambda-types.js` 🎉

File: package.json
Comment (corinja) [reply package.json:53]: > Suggesting we update to the latest version of `axios` before merging.

Latest version of axios doesn't seem to play nicely with jest :(

Comment (corinja) [reply package.json:53]: https://github.com/axios/axios/issues/5101

Comment (pauldolden) [reply package.json:53]: @corinja currently working on #73, which migrates this repo to TypeScript as per the discussion in the planning session yesterday. I've made some changes to the jest setup here so it may well be the case that it will be fine. This PR is most of the way there, and is largely just waiting on this PR to be merged.

I can port these changes over to my PR and see if there is any issue with the axios version if that is a suitable suggestion?

Comment (corinja) [reply package.json:53]: > Hey @corinja, I'm currently working on #73, which migrates this repo to TypeScript as per the discussion in the planning session yesterday. I've made some changes to the jest setup here so it may well be the case that it will be fine. This PR is most of the way there, and is largely just waiting on this PR to be merged.
> 
> I can port these changes over to my PR and see if there is any issue with the axios version if that is a suitable suggestion?

I didn't see your comment before posting my update @pauldolden I'm just about to push my changes FYI

File: tests/feat/hello.spec.js
Comment (corinja) [reply tests/feat/hello.spec.js:12]: Sure. I am going to `lower-case` the keys in the header as well, since AWS gives them in proper case but sls offline uses all lower case, it seems.

File: tests/lib/actions/hello.js
Comment (corinja) [reply tests/lib/actions/hello.js]: Done this

File: tests/feat/hello.spec.js
Comment (corinja) [reply tests/feat/hello.spec.js]: Done

Comment (corinja) [reply tests/feat/hello.spec.js]: There might be a better way to `lowerCase` the header keys - I'm open to suggestion!

File: package.json
Comment (corinja) [reply package.json]: Done!

File: tests/feat/hello.spec.js
Comment (seb-cr) [reply tests/feat/hello.spec.js]: The header cases are different, but `axios` already deals with this 🙂 

[See docs](https://github.com/axios/axios#response-schema)

> All header names are lowercase and can be accessed using the bracket notation.

Comment (corinja) [reply tests/feat/hello.spec.js]: Ah, ok - have removed my bodge

**Comment by corinja** [comment]
Do you think this is ok to merge now @seb-cr ?

**Review by seb-cr — COMMENTED** [review]
Had a quick read through again and spotted a couple of minor things related to the last batch of changes. Everything else looks good, thanks.

File: package.json
Comment (corinja) [reply package.json:53]: @seb-cr is there a way to do this using `yarn`?

File: tests/setup.js
Comment (corinja) [reply tests/setup.js]: I've removed the comment completely, didn't add anything

File: package.json
Comment (seb-cr) [reply package.json:53]: `yarn add axios` should do it automatically, or failing that, `yarn add axios@^1.3.6`

File: tests/setup.js
Comment (seb-cr) [reply tests/setup.js]: Cool, agree it's pretty self-explanatory.

**Review by seb-cr — APPROVED** [review]

**Pull request merged**
